### PR TITLE
Combined reco

### PR DESCRIPTION
**Pull request by faltovaj**
This PR enables running the calorimeter reconstruction (sliding window algorithm) on all calorimeter systems (Ecal barrel, Hcal barrel + extended barrel, Endcaps and Forward). Moreover, it includes updates on the segmentation in the Calo endcaps & forward. 

Changes proposed in this pull-request:
- Clustering algorithm running on the full calorimeter system. SingleCaloTowerTool removed, CombinedCaloTowerTool to be used in all cases. If some of the cell collections are missing, use CreateDummyCellsCollection to create an empty collection of calo cells.
- Segmentation in the Calo endcaps & forward changed to match our baseline cell sizes
- Number of tests and scripts removed, as they are merged in the full calorimeter reconstruction example script

### CODE REVIEW

**Comment by CoralieNeubueser** [comment]
Hi Jana,
thanks for this important addition of the tower tool and the clean-up of the simulation scripts!
I have one general question, is the creation of these empty dummy collections really necessary?
I mean isn't it possible to simply ask for a collection and and if it is not given, ignored?
Thanks,
Coralie

File: Detector/DetComponents/src/MergeLayers.cpp [hunk 1]
```diff
@@ -55,16 +55,6 @@ StatusCode MergeLayers::initialize() {
             << endmsg;
     return StatusCode::FAILURE;
   }
-  // check sizes of new volumes in the list - it must sum to the total number of volumes
```
Comment (CoralieNeubueser) [inline Detector/DetComponents/src/MergeLayers.cpp]: Why did you remove this check?

File: Reconstruction/RecCalorimeter/tests/options/runFullCaloSystem_SimAndDigitisation.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+from Configurables import CalibrateInLayersTool
+calibEcalBarrel = CalibrateInLayersTool("CalibrateECalBarrel",
+                                   # sampling fraction obtained using SamplingFractionInLayers from DetStudies package
+                                   samplingFraction = [0.12125] + [0.14283] + [0.16354] + [0.17662] + [0.18867] + [0.19890] + [0.20637] + [0.20802],
+                                   readoutName = ecalBarrelReadoutName,
+                                   layerFieldName = "layer")
+calibEndcap = CalibrateInLayersTool("CalibrateEndcap",
+                                    # sampling fraction obtained using SamplingFractionInLayers from DetStudies package
+                                    # HEC -> extrapolation from Ecal EC (0.15 * active / passive thickness)
+                                    samplingFraction = [0.15] * 118 + [0.03] * 78,
```
Comment (CoralieNeubueser) [inline Reconstruction/RecCalorimeter/tests/options/runFullCaloSystem_SimAndDigitisation.py]: The calibration is applied to the readout of the EMEC only, isn't it?
For the calibration of the HEC another calibration tool has do be used, hasn't it?

File: Reconstruction/RecCalorimeter/tests/options/runFullCaloSystem_SimAndDigitisation.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+                                # new bitfield (readout), with new segmentation
+                                newReadoutName = newExtHcalReadoutName,
+                                debugPrint = 10,
+                                OutputLevel = INFO,
+                                inhits = "ExtHCalPositions",
+                                outhits = "newExtHCalCells")
+
+out = PodioOutput("out", 
+                  OutputLevel=INFO)
+out.outputCommands = ["keep *", "drop ECalBarrelHits", "drop ECalBarrelPositionedHits", "drop HCalHits", "drop HCalPositionedHits", "drop HCalCells", "drop ExtHCalHits", "drop ExtHCalPositionedHits", "drop ExtHCalCells", "drop ECalBarrelCellsStep1", "drop CalEndcapHits", "drop CalEndcapPositionedHits", "drop CalFwdHits", "drop CalFwdPositionedHits"]
```
Comment (zaborowska) [inline Reconstruction/RecCalorimeter/tests/options/runFullCaloSystem_SimAndDigitisation.py]: maybe it is easier (and quicker) to say what do you want to keep?

File: Detector/DetComponents/src/MergeLayers.cpp (hunk 1 above)
Comment (faltovaj) [reply Detector/DetComponents/src/MergeLayers.cpp]: It was causing troubles if the same volume name is used in different systems. We have "layer" in the ECAL and the same name is used also in some other subdetector. After discussion with Anna, the author of this tool, we decided to remove this check and leave the responsibility on the user of the code.

File: Reconstruction/RecCalorimeter/tests/options/runFullCaloSystem_SimAndDigitisation.py (hunk 2 above)
Comment (faltovaj) [reply Reconstruction/RecCalorimeter/tests/options/runFullCaloSystem_SimAndDigitisation.py]: The calibration is done using 2 different sampling fractions, one for ECAL part and the second one for the HCAL part. The calibration for HEC and forward hadronic calo was not calculated properly, it is a first guess. I extrapolated the numbers from the EMEC and EFCAL calibrations taking into account different thicknesses of the passive/active material.

**Comment by faltovaj** [comment]
Hi @CoralieNeubueser ,
thanks for looking into the PR and checking the changes.
Unfortunately, it is not possible to ignore a collection. The simplest solution we found was with the creation of the dummy cells collection.

File: Reconstruction/RecCalorimeter/tests/options/runFullCaloSystem_SimAndDigitisation.py (hunk 2 above)
Comment (CoralieNeubueser) [reply Reconstruction/RecCalorimeter/tests/options/runFullCaloSystem_SimAndDigitisation.py]: That I understood, but the readout is done for the EM and hadronic parts seperately, isn't it?

File: Detector/DetCommon/src/DetUtils.cpp [hunk 4]
```diff
@@ -140,14 +140,16 @@ CLHEP::Hep3Vector tubeDimensions(uint64_t aVolumeId) {
...
   // get segmentation number of bins in phi
   uint phiCellNumber = aSeg.phiBins();
   // get segmentation cell width in eta
   double etaCellSize = aSeg.gridSizeEta();
-  // eta segmentation calculate maximum eta from the inner radius (no offset is taken into account)
-  double maxEta = CLHEP::Hep3Vector(tubeSizes.x(), 0, tubeSizes.z()).eta();
-  uint cellsEta = ceil((maxEta - etaCellSize / 2.) / etaCellSize) * 2 + 1;
+  // eta segmentation calculate maximum eta from the inner radius, offset is taken into account
+  double maxEta = CLHEP::Hep3Vector(tubeSizes.x(), 0, tubeSizes.z()).eta() + fabs(aSeg.offsetEta() - etaCellSize * 0.5);
+  double minEta = - CLHEP::Hep3Vector(tubeSizes.x(), 0, tubeSizes.z()).eta() + fabs(aSeg.offsetEta() - etaCellSize * 0.5);
```
Comment (CoralieNeubueser) [inline Detector/DetCommon/src/DetUtils.cpp]: I think it should be: 
double minEta = - (CLHEP::Hep3Vector(tubeSizes.x(), 0, tubeSizes.z()).eta() + fabs(aSeg.offsetEta() - etaCellSize * 0.5));

File: Detector/DetFCChhECalInclined/compact/FCChh_ECalBarrel_Common.xml [hunk 5]
```diff
@@ -60,7 +60,7 @@
     <!-- phi position is calculated based on the centre of volume (hence it cannot be done in the simulation from energy deposits position) -->
     <readout name="ECalBarrelPhiEta">
       <segmentation type="GridPhiEta" grid_size_eta="0.01" phi_bins="704" offset_eta="-1.68024" offset_phi="-pi+(pi/704.)"/>
-      <id>system:4,cryo:1,module:11,type:3,subtype:3,layer:8,eta:9,phi:10</id>
```
Comment (CoralieNeubueser) [inline Detector/DetFCChhECalInclined/compact/FCChh_ECalBarrel_Common.xml]: Hi Jana, 
I merged with your branch and get some issues with the matching between CellID and volumeID (for the retrieval of cell positions) because of the removed "module" field in the ECAL readout. Could we keep this in, or do you have a strong reason for rm it?

**Comment by faltovaj** [comment]
Dear @CoralieNeubueser and @zaborowska,
thanks a lot for having a look into the PR. I hope that all the changes required are in now.
A couple of comments:
1/ Finally I decided to remove the changes in the DetUtils where I wanted to take the offset into account. It is working fine as it is for the barrel. We cannot use the same for the endcaps, as it is not a tube geometry anymore. It will not help in this case.
2/ I left the "module" ID in the inclined Ecal readout, as discussed with Coralie.
Let me know if you have any other comments.
Cheers,
Jana

File: Detector/DetCommon/src/DetUtils.cpp [hunk 6]
```diff
@@ -145,6 +145,12 @@ std::array<uint, 2> numberOfCells(uint64_t aVolumeId, const DD4hep::DDSegmentati
   uint phiCellNumber = aSeg.phiBins();
   // get segmentation cell width in eta
   double etaCellSize = aSeg.gridSizeEta();
+  // eta segmentation calculate maximum eta from the inner radius, offset is taken into account
+  /*
```
Comment (zaborowska) [inline Detector/DetCommon/src/DetUtils.cpp]: Does that mean we no longer take offset into account?
Also, commented code should be removed.

File: Reconstruction/RecCalorimeter/src/components/CalibrateInLayersTool.cpp [hunk 7]
```diff
@@ -38,7 +38,9 @@ void CalibrateInLayersTool::calibrate(std::unordered_map<uint64_t, double>& aHit
     } else {
       p.second /= m_samplingFraction[m_samplingFraction.size() - 1];
       warning() << "Size of sampling fraction values is smaller than the number of existing layers."
-                << " Taking the sampling fraction for last layer." << endmsg;
+                << " Taking the sampling fraction for last layer." 
+		<< " Layer ID: " << layer
```
Comment (zaborowska) [inline Reconstruction/RecCalorimeter/src/components/CalibrateInLayersTool.cpp]: Fix indentation (checkformat)

File: Reconstruction/RecCalorimeter/tests/options/runFullCaloSystem_SimAndDigitisation.py [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+
+from Configurables import CalibrateInLayersTool
+calibEcalBarrel = CalibrateInLayersTool("CalibrateECalBarrel",
+                                   # sampling fraction obtained using SamplingFractionInLayers from DetStudies package
+                                   samplingFraction = [0.12125] + [0.14283] + [0.16354] + [0.17662] + [0.18867] + [0.19890] + [0.20637] + [0.20802],
+                                   readoutName = ecalBarrelReadoutName,
+                                   layerFieldName = "layer")
+calibEcalEndcap = CalibrateInLayersTool("CalibrateEcalEndcap",
+                                    # sampling fraction obtained using SamplingFractionInLayers from DetStudies package
+                                    samplingFraction = [0.15] * 119,
```
Comment (zaborowska) [inline Reconstruction/RecCalorimeter/tests/options/runFullCaloSystem_SimAndDigitisation.py]: can you correct to the new calib constants?

File: Reconstruction/RecCalorimeter/src/components/SingleCaloTowerTool.cpp [hunk 9]
```diff
@@ -1,183 +0,0 @@
-#include "SingleCaloTowerTool.h"
```
Comment (zaborowska) [inline Reconstruction/RecCalorimeter/src/components/SingleCaloTowerTool.cpp]: Do we want to remove completely this class? It's not going to be useful e.g. for debugging? Or simple tests? (Do you think it's going to create problems to keep both parts updated).
Because if so maybe the other class should be called simply CaloTowerTool (without Combined)? - just an idea

File: Reconstruction/RecCalorimeter/tests/options/runFullCaloSystem_SimAndDigitisation.py (hunk 8 above)
Comment (faltovaj) [reply Reconstruction/RecCalorimeter/tests/options/runFullCaloSystem_SimAndDigitisation.py]: It is done using CalibrateCaloHitsTool, as the sampling fraction is the same for all layers. Just note that the 1/sampling fraction is the parameter in this tool.

File: Reconstruction/RecCalorimeter/src/components/SingleCaloTowerTool.cpp (hunk 9 above)
Comment (faltovaj) [reply Reconstruction/RecCalorimeter/src/components/SingleCaloTowerTool.cpp]: Yes, I prefer to completely remove this tool. It is quite straight forward to use the general tool with only one collection as it is done in runEcalSimple_ReconstructionSW_noiseFromFile.py. I renamed the tool as you suggested.

File: Detector/DetCommon/src/DetUtils.cpp (hunk 6 above)
Comment (faltovaj) [reply Detector/DetCommon/src/DetUtils.cpp]: I decided to remove my changes, as it will not calculate the number of cells in the endcaps or forward because of their special shape. I'd propose to leave it as it was and think about it once more in the next PR.

File: Reconstruction/RecCalorimeter/src/components/SingleCaloTowerTool.cpp (hunk 9 above)
Comment (zaborowska) [reply Reconstruction/RecCalorimeter/src/components/SingleCaloTowerTool.cpp]: ok, thanks!

File: Reconstruction/RecCalorimeter/tests/options/runFullCaloSystem_SimAndDigitisation.py (hunk 8 above)
Comment (zaborowska) [reply Reconstruction/RecCalorimeter/tests/options/runFullCaloSystem_SimAndDigitisation.py]: Sorry, I added this comment before you pushed those changes. Indeed it is ok.

**Comment by zaborowska** [comment]
retest this please

**Comment by faltovaj** [comment]
retest this please

**Comment by hegner** [comment]
retest this please

**Comment by faltovaj** [comment]
retest this please

**Comment by zaborowska** [comment]
retest this please

**Comment by zaborowska** [comment]
retest this please

**Comment by hegner** [comment]
can you change the debug level of the failing test to something less verbose than INFO?

**Comment by zaborowska** [comment]
retest this please

**Comment by zaborowska** [comment]
@faltovaj can you fix the conflict in combine simulation options? The inverse sampling fraction changed for tile HCal with Pb spacers to 41.66.
